### PR TITLE
feat: [NPM] define transport shim layer using gRPC

### PR DESCRIPTION
**Pull request by nitishm**
User story AB#12707964

**Reason for Change**:
Define protobuf message and services for transport shim

**Issue Fixed**:
AB#12707965

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

Tasks:

- [X] Define protobuf file for all Datapath interface methods

Signed-off-by: Nitish Malhotra <nitishm@microsoft.com>

**Notes**:

### CODE REVIEW

**Comment by rbtr** [comment]
how are you generating the pb.go? would be nice to define gen targets in the makefile for consistency 🙂 
even better if you also don't assume protoc is installed and take care of that in a similar fashion as the existing go tools dependency targets.

**Comment by nitishm** [comment]
> how are you generating the pb.go? would be nice to define gen targets in the makefile for consistency 🙂 even better if you also don't assume protoc is installed and take care of that in a similar fashion as the existing go tools dependency targets.

Sure I can do that. Are we using it any place else today? I did see a few proto files across the code base.

**Comment by vakalapa** [comment]
> > how are you generating the pb.go? would be nice to define gen targets in the makefile for consistency 🙂 even better if you also don't assume protoc is installed and take care of that in a similar fashion as the existing go tools dependency targets.
> 
> Sure I can do that. Are we using it any place else today? I did see a few proto files across the code base.

@nitishm  https://github.com/Azure/azure-container-networking/blob/master/npm/pkg/dataplane/Makefile Check this, Mat has created this MakeFile for auto-generating go mock. You can do something similar. My opinion is this does not have to be in global MakeFile, since we will only generate this once in a while.

File: scripts/install-protoc.sh
Comment (rbtr) [inline scripts/install-protoc.sh:29]: can you set up your editor to add a trailing newline on save?

File: Makefile
Comment (rbtr) [inline Makefile:429]: we don't have a well-defined project layout, but should this scripts dir be under `/build` maybe?

Comment (nitishm) [reply Makefile]: From what I have observed most scripts go into $ROOT/scripts (or at least in the kubebuilder project and other go projects). But if build seems like the right place for all scripts then we should definitely move it there. 
@vakalapa chime in?

File: scripts/install-protoc.sh
Comment (nitishm) [reply scripts/install-protoc.sh:45]: Oh - how do I do that? I am using vscode

**Comment by nitishm** [comment]
To test this out use the following instructions:

Download the protoc binary
```terminal
./scripts/install-protoc.sh 
```

Install the grpc go gen tool
```terminal
make tools
```

Run generate target
```terminal
cd npm
make generate
```

File: scripts/install-protoc.sh
Comment (rbtr) [reply scripts/install-protoc.sh:45]: Settings > Files: Insert Final Newline

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [inline npm/pkg/transport/pb/transport.proto:15]: Do we need a separate ID ? because we can use Node name as the ID.

**Review by vakalapa — APPROVED** [review]
Approving with minor comments.

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:15]: I think pod name would be good to track (for future use probably)?

**Comment by nitishm** [comment]
@vakalapa can I get another approval. Also any idea why the CI doesn't like it?

File: npm/pkg/transport/pb/transport.proto
Comment (ramiro-gamarra) [inline npm/pkg/transport/pb/transport.proto:3]: For package names, it might be better to communicate domain. In this case `pb` seems a bit generic and it might clash with other packages. Thoughts `dataplane` or `networkpolicy`? Seems like that is the domain of the services and types below.

File: npm/pkg/transport/pb/transport.proto
Comment (ramiro-gamarra) [reply npm/pkg/transport/pb/transport.proto:15]: nit: maybe better to be very explicit about the identifiers? `pod_id` and `node_name` seem less ambiguous to me.

File: npm/pkg/transport/pb/transport.proto
Comment (ramiro-gamarra) [inline npm/pkg/transport/pb/transport.proto:46]: I'm guessing this is still in prototyping stages, but I've usually found working with `Struct` difficult. Is the structure of IPSet and NetworkPolicy well defined? Could use `oneof`

**Review by ramiro-gamarra — COMMENTED** [review]
Got curious about this, so I'm leaving a few comments/questions :)

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:46]: I am afraid that we might get out of sync if things change where someone might forget to update the proto file

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:15]: sure I can do that

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:3]: Since it is local to the transport package I don't think it should matter

File: npm/pkg/transport/pb/transport.proto
Comment (rbtr) [reply npm/pkg/transport/pb/transport.proto:46]: make updating the proto part of the build and fail if it diffs dirty

Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:46]: How do you diff it with a standard go struct? The proto generated structure will most likely vary from the existing objects due to lack of type support.

Comment (ramiro-gamarra) [reply npm/pkg/transport/pb/transport.proto:46]: you would diff what the build generates vs what's checked in. this ensures that go proto code is up to date with the source of truth, the proto file.

Comment (ramiro-gamarra) [reply npm/pkg/transport/pb/transport.proto:46]: I still don't think I understand how `Struct` helps in this case. For prototyping purposes it should be fine, but it's a lot easier to for client and server to get out of sync when using it since it is basically like json and foregoes the benefits of strong typing.

Comment (rbtr) [reply npm/pkg/transport/pb/transport.proto:46]: > diff what the build generates vs what's checked in

to expand, you `make gen` and then something like `test -z "$(git status --porcelain)"` which returns 0 if the pwd is clean and 1 if there are changes vs source

Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:46]: > you would diff what the build generates vs what's checked in. this ensures that go proto code is up to date with the source of truth, the proto file.

I don't mean the proto generated code. I mean the actual IPSet, and other structs defined in `types.go`.

Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:46]: > I still don't think I understand how `Struct` helps in this case. For prototyping purposes it should be fine, but it's a lot easier to for client and server to get out of sync when using it since it is basically like json and foregoes the benefits of strong typing.

I agree. This is not set in stone. Good for testing. I will play around with it and figure out what works best.

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [reply npm/pkg/transport/pb/transport.proto:15]: Also, do we want to introduce an "api_version" field to help us maintain the structure in long run for upgrade cases ? Both for PodMetadata and also events

Comment (vakalapa) [reply npm/pkg/transport/pb/transport.proto:15]: <Nit> we normally use "daemonPod" instead of DataplanePod, helps us with consistent naming.

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [inline npm/pkg/transport/pb/transport.proto:32]: <nit> remove "rules" like below 

>// Supported objects are IPSet and NetworkPolicy objects.

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [inline npm/pkg/transport/pb/transport.proto:44]: <nit> container -> contain ?

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [reply npm/pkg/transport/pb/transport.proto:46]: @nitishm  once we build a goatStateProcessor with the type generated from this proto file then if proto generation gets dirty, GSProcessor build should fail (ideally). Lets move forward at this moment with current "struct" design, this is a 2-way door, as we build bits around this, i am sure we will tweak this struct further, so it should be okay with current type. Wdygt ?

File: npm/pkg/transport/transport.go
Comment (vakalapa) [inline npm/pkg/transport/transport.go:72]: there can some potential race conditions here ?  what if a connection got lost and created for the same pod in immediate succession, and lets say deregCh is overloaded and regCh is free, then there is a high chance that we would delete a legitimate connection right ? how do we manage those kinds of race conditions ?

We might need to introduce an aspect of timestamp ?

File: npm/pkg/transport/transport.go
Comment (vakalapa) [inline npm/pkg/transport/transport.go:15]: We might have to think about certs management in Manager too ?

**Review by vakalapa — COMMENTED** [review]
Minor comments, please resolve all lints.

File: npm/pkg/transport/transport.go
Comment (nitishm) [reply npm/pkg/transport/transport.go:72]: I think a ping-ping health check similar to Felix and typha is a must to avoid such issues. We need some keepalives and checks in place. Thoughts?

Comment (nitishm) [reply npm/pkg/transport/transport.go:72]: Yea I think timestamps makes sense.

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:15]: Sticking to Pod and Node to keep it generic. I don't understand why the file is showing the older version 
We have `pod` and `node` as the two fields

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [inline npm/pkg/transport/pb/transport.proto:16]: @vakalapa making this an enum so that we default to version 1 if not specified. Easier to track versions with enum. WDYT?

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:15]: > nit: maybe better to be very explicit about the identifiers? `pod_id` and `node_name` seem less ambiguous to me.

Changed it to `pod` and `node` to keep it simple. The id suffix is redundant IMO. WDYT?

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:3]: @ramiro-gamarra Thoughts?

File: npm/pkg/transport/pb/transport.proto
Comment (ramiro-gamarra) [reply npm/pkg/transport/pb/transport.proto:15]: Don't let my feedback on this be a blocker for prototyping purposes btw :), just sharing my 2 cents. Thing about contracts is to keep ambiguity to a minimum IMO. `pod` could be understood as the name of the pod, or the uid of the pod, 2 different things. Maybe documentation could help in the proto file itself, but when used in the generated code, it would also have a bit of ambiguity.

File: npm/pkg/transport/pb/transport.proto
Comment (ramiro-gamarra) [reply npm/pkg/transport/pb/transport.proto:3]: Not sure I understand the part about local. The imported package would still be `pb` and referenced as such across the code. If the client had a second grpc service to interact with for example, and its package was under `pkg/foo/pb`, it would also be imported as `pb` but would conflict with the other one, even if they are unrelated paths. Maybe I'm missing a detail, let me know

File: npm/pkg/transport/transport.go
Comment (ramiro-gamarra) [inline npm/pkg/transport/transport.go:94]: should probably return here, otherwise the rest of the goroutine executes with a nil listener after the error is sent. maybe it would be better to have the grpc server wired synchronously and then spinning `server.Server` in a goroutine?

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:3]: I see your point now. @vakalapa thoughts? Since this proto is common across both cp and dp how would we name it
Maybe we restructure as we get it it? @vakalapa what are your thoughts on this?

File: npm/pkg/transport/transport.go
Comment (nitishm) [reply npm/pkg/transport/transport.go:94]: Yup that works. Making the change now.

Comment (nitishm) [reply npm/pkg/transport/transport.go]: Done - PTAL @ramiro-gamarra

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [reply npm/pkg/transport/pb/transport.proto:3]: Ack, i can see that confusion arising, and we will be expanding later on, so having a specific pkg name instead of pb would help for sure, 

Other option is to consolidate all protos into a central location under like below (calico does this) so all other pks can access these types without any cyclic issues 

>npm/pkg/protos

File: npm/pkg/transport/pb/transport.proto
Comment (vakalapa) [reply npm/pkg/transport/pb/transport.proto:16]: that makes sense!

File: npm/pkg/transport/transport.go
Comment (nitishm) [inline npm/pkg/transport/transport.go:125]: I cant recall if closures would prevent the member channel from being passed as-is into the go routine's scope or not. So I explicitly pass it in as a safety measure 🤷

Comment (nitishm) [inline npm/pkg/transport/transport.go:125]: I cant recall if closures would prevent the member channel from being passed as-is into the go routine's scope or not. So I explicitly pass it in as a safety measure 🤷

Comment (rbtr) [reply npm/pkg/transport/transport.go:125]: as long as it's not reassigned in a loop, the variable gets captured in the closure

Comment (rbtr) [reply npm/pkg/transport/transport.go:125]: [this](https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables) is probably what you're thinking of?

Comment (nitishm) [reply npm/pkg/transport/transport.go:125]: Ye I knew about that one but wasn't sure if channels might have a similar fate. Seems unlikely

File: npm/pkg/transport/consts.go
Comment (ramiro-gamarra) [inline npm/pkg/transport/consts.go:6]: seems like this is used as a key to fetch context values. usually these are their own unexported types in order to avoid key collision in the global scope.

Comment (nitishm) [reply npm/pkg/transport/consts.go]: Done.

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:15]: Sure I can update that to `pod_name` & `node_name`

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto:3]: Let me see what I can do

Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto]: Done

File: npm/pkg/transport/pb/transport.proto
Comment (nitishm) [reply npm/pkg/transport/pb/transport.proto]: Done

**Review by vakalapa — APPROVED** [review]

**Pull request merged**
